core/trino-main/src/main/java/io/trino/sql/planner/RuntimeAdaptivePartitioningRewriter.java

```diff
@@ -0,0 +1,217 @@
+                    fragment,
+                    subPlan.getChildren().stream()
+                            .map(child -> overridePartitionCountRecursively(
+                                    child,
+                                    oldPartitionCount,
+                                    newPartitionCount,
+                                    planFragmentIdAllocator,
+                                    planNodeIdAllocator,
+                                    startedFragments))
+                            .collect(toImmutableList()));
+        }
+
+        // insert extra exchanges to sources
+        ImmutableList.Builder<SubPlan> newSources = ImmutableList.builder();
+        ImmutableMap.Builder<PlanFragmentId, PlanFragmentId> runtimeAdaptivePlanFragmentIdMapping = ImmutableMap.builder();
+        for (SubPlan source : subPlan.getChildren()) {
+            PlanFragment sourceFragment = source.getFragment();
+            RemoteSourceNode sourceRemoteSourceNode = getOnlyElement(fragment.getRemoteSourceNodes().stream()
+                    .filter(remoteSourceNode -> remoteSourceNode.getSourceFragmentIds().contains(sourceFragment.getId()))
+                    .iterator());
+            requireNonNull(sourceRemoteSourceNode, "sourceRemoteSourceNode is null");
+            if (sourceRemoteSourceNode.getExchangeType() == REPLICATE) {
+                // no need to insert extra exchanges
+                // since exchange type is REPLICATE, also no need to change partition count for the subtree
+                newSources.add(source);
+                continue;
+            }
+            if (!startedFragments.contains(sourceFragment.getId())) {
+                // source not started yet, then no need to insert extra exchanges to sources
+                newSources.add(overridePartitionCountRecursively(
+                        source,
+                        oldPartitionCount,
+                        newPartitionCount,
+                        planFragmentIdAllocator,
+                        planNodeIdAllocator,
+                        startedFragments));
+                runtimeAdaptivePlanFragmentIdMapping.put(sourceFragment.getId(), sourceFragment.getId());
+                continue;
+            }
+            RemoteSourceNode runtimeAdaptiveRemoteSourceNode = new RemoteSourceNode(
+                    planNodeIdAllocator.getNextId(),
+                    sourceFragment.getId(),
+                    sourceFragment.getOutputPartitioningScheme().getOutputLayout(),
+                    sourceRemoteSourceNode.getOrderingScheme(),
+                    sourceRemoteSourceNode.getExchangeType(),
+                    sourceRemoteSourceNode.getRetryPolicy());
+            PlanFragment runtimeAdaptivePlanFragment = new PlanFragment(
+                    planFragmentIdAllocator.getNextId(),
+                    runtimeAdaptiveRemoteSourceNode,
+                    sourceFragment.getSymbols(),
+                    FIXED_HASH_DISTRIBUTION,
```

Comment:
Why do you think it will cause a problem? At the end, the stage with `SCALED_WRITER_HASH_DISTRIBUTION` will consume the output of the new stage, so I don't see how things might break. Maybe it's because I'm not familiar with scaled writers.

@arhimondr , thoughts?

Reply:
Discussed offline, it's not possible we will add a new PlanFragment as the parent of a `SCALED_WRITER_HASH_DISTRIBUTION` fragment